### PR TITLE
st1 [1782][IMP] stock_delivery_result_import:prohibited update if carrier is blank

### PR DESCRIPTION
**Pull request by TomohiroAbe-Quartile**
[1782](https://www.quartile.co/web#id=1782&action=771&model=project.task&view_type=form&menu_id=505)

### CODE REVIEW

File: stock_delivery_result_import/wizards/stock_delivery_result_import.py [hunk 1]
```diff
@@ -72,16 +72,24 @@ def import_stock_delivery_result(self):
                 )
         if not import_log.error_ids:
             for picking, vals in pick_dict.items():
-                carrier_info = vals["carrier_info"]
+                if len(carrier_info) != 0:
+                    carrier_info_val = vals["carrier_info"]
+                else:
+                    carrier_info_val = False
```
Comment (yostashiro) [inline stock_delivery_result_import/wizards/stock_delivery_result_import.py]: @TomohiroAbe-Quartile Not getting what you are trying to do here.  Can you first describe what issues you observe from debugging?

Comment (TomohiroAbe-Quartile) [reply stock_delivery_result_import/wizards/stock_delivery_result_import.py]: @yostashiro 
Sorry, I cannot debug in my dev environment.

Error occurs and debug couldn't continue like below:

https://github.com/odoo/odoo/blob/28c3f51c4878fbcd79b2e819948465fcf2160ebc/odoo/modules/registry.py#L177

```Exception has occurred: KeyError       (note: full exception trace is shown but execution is paused at: __getitem__)
(None,)
  File "/home/tomohiro/ws/mi7/odoo/odoo/modules/registry.py", line 177, in __getitem__
    return self.models[model_name]
  File "/home/tomohiro/ws/mi7/odoo/odoo/api.py", line 760, in __getitem__ (Current frame)
    return self.registry[model_name]._browse((), self)
  File "/home/tomohiro/ws/mi7/odoo/odoo/fields.py", line 739, in resolve_deps
    model = model0.env.get(field.comodel_name)
  File "/home/tomohiro/ws/mi7/odoo/odoo/fields.py", line 763, in setup_triggers
    for model, field, path in self.resolve_deps(model):
  File "/home/tomohiro/ws/mi7/odoo/odoo/models.py", line 2896, in _setup_complete
    field.setup_triggers(self)
  File "/home/tomohiro/ws/mi7/odoo/odoo/modules/registry.py", line 303, in setup_models
    model._setup_complete()
  File "/home/tomohiro/ws/mi7/odoo/odoo/modules/loading.py", line 135, in load_module_graph
    registry.setup_models(cr, partial=True)
  File "/home/tomohiro/ws/mi7/odoo/odoo/modules/loading.py", line 270, in load_marked_modules
    perform_checks=perform_checks, models_to_check=models_to_check
  File "/home/tomohiro/ws/mi7/odoo/odoo/modules/loading.py", line 375, in load_modules
    force, status, report, loaded_modules, update_module, models_to_check)
  File "/home/tomohiro/ws/mi7/odoo/odoo/modules/registry.py", line 83, in new
    odoo.modules.load_modules(registry._db, force_demo, status, update_module)
  File "/home/tomohiro/ws/mi7/odoo/odoo/modules/registry.py", line 59, in __new__
    return cls.new(db_name)
  File "/home/tomohiro/ws/mi7/odoo/odoo/__init__.py", line 52, in registry
    return modules.registry.Registry(database_name)
  File "/home/tomohiro/ws/mi7/odoo/odoo/http.py", line 1482, in dispatch
    odoo.registry(db).check_signaling()
  File "/home/tomohiro/ws/mi7/odoo/odoo/http.py", line 1300, in __call__
    return self.app(environ, start_wrapped)
  File "/home/tomohiro/ws/mi7/odoo/odoo/http.py", line 1326, in __call__
    return self.dispatch(environ, start_response)
  File "/home/tomohiro/ws/mi7/odoo/odoo/service/wsgi_server.py", line 172, in application_unproxied
    result = handler(environ, start_response)
  File "/home/tomohiro/ws/mi7/odoo/odoo/service/wsgi_server.py", line 186, in application
    return application_unproxied(environ, start_response)
  File "/home/tomohiro/ws/mi7/odoo/odoo/service/server.py", line 307, in app
    return self.app(e, s)```

File: stock_delivery_result_import/wizards/stock_delivery_result_import.py [hunk 2]
```diff
@@ -74,14 +76,16 @@ def import_stock_delivery_result(self):
...
-                        "carrier_info_id": carrier_info and carrier_info.id or False,
-                        "carrier_tracking_ref": tracking_refs,
-                        "log_id": import_log.id,
-                    }
+                    {"carrier_tracking_ref": tracking_refs, "log_id": import_log.id}
                 )
+                if not picking.carrier_info_id:
+                    picking.write(
+                        {"carrier_info_id": carrier_info and carrier_info.id or False}
+                    )
```
Comment (yostashiro) [inline stock_delivery_result_import/wizards/stock_delivery_result_import.py]: @TomohiroAbe-Quartile Avoid splitting the write operation on picking in two separate steps unless it's necessary.

Comment (TomohiroAbe-Quartile) [reply stock_delivery_result_import/wizards/stock_delivery_result_import.py]: @yostashiro 
Sorry, I have no idea why unnecessary to splitting the write operation in this situation.

Comment (TomohiroAbe-Quartile) [reply stock_delivery_result_import/wizards/stock_delivery_result_import.py]: I edit carrier_info define and now it works as expected.

File: stock_delivery_result_import/wizards/stock_delivery_result_import.py [hunk 3]
```diff
@@ -60,7 +60,9 @@ def import_stock_delivery_result(self):
                             "tracking_refs": [],
                             "carrier_info": carrier_info,
                         }
-                pick_dict[picking]["tracking_refs"].append(row_dict.get("tracking_ref"))
+                        pick_dict[picking]["tracking_refs"].append(
+                            row_dict.get("tracking_ref")
+                        )
```
Comment (yostashiro) [inline stock_delivery_result_import/wizards/stock_delivery_result_import.py]: Why this change?  It will make the code capture only the first tracking ref if there are multiple tracking refs involved.

File: stock_delivery_result_import/wizards/stock_delivery_result_import.py [hunk 4]
```diff
@@ -72,14 +74,20 @@ def import_stock_delivery_result(self):
...
             for picking, vals in pick_dict.items():
-                carrier_info = vals["carrier_info"]
+                carrier_info = (
+                    picking.carrier_info_id
+                    if picking.carrier_info_id
+                    else vals["carrier_info"]
+                )
                 tracking_refs = list(set(vals["tracking_refs"]))
+                if picking.carrier_tracking_ref:
+                    tracking_refs.append(picking.carrier_tracking_ref)
```
Comment (yostashiro) [inline stock_delivery_result_import/wizards/stock_delivery_result_import.py]: This manipulation should be done here https://github.com/qrtl/mi7-custom/pull/87/files#diff-037e9962dee26567028229c7b0c134ac871bc72397305b6865a6ad61f2d2413eL60 instead.

File: stock_delivery_result_import/wizards/stock_delivery_result_import.py [hunk 5]
```diff
@@ -72,14 +74,20 @@ def import_stock_delivery_result(self):
                 )
         if not import_log.error_ids:
             for picking, vals in pick_dict.items():
-                carrier_info = vals["carrier_info"]
+                carrier_info = (
+                    picking.carrier_info_id
+                    if picking.carrier_info_id
+                    else vals["carrier_info"]
+                )
```
Comment (yostashiro) [inline stock_delivery_result_import/wizards/stock_delivery_result_import.py]: This manipulation should be done here https://github.com/qrtl/mi7-custom/pull/87/files#diff-037e9962dee26567028229c7b0c134ac871bc72397305b6865a6ad61f2d2413eL56-L58 instead.

File: stock_delivery_result_import/wizards/stock_delivery_result_import.py [hunk 6]
```diff
@@ -72,14 +74,20 @@ def import_stock_delivery_result(self):
...
                 tracking_refs = list(set(vals["tracking_refs"]))
+                if picking.carrier_tracking_ref:
+                    tracking_refs.append(picking.carrier_tracking_ref)
                 tracking_refs = ", ".join(ref for ref in tracking_refs)
                 picking.write(
```
Comment (yostashiro) [inline stock_delivery_result_import/wizards/stock_delivery_result_import.py]: Update `vals` for the available variables only first, and then do `picking.write(vals)`.

**Review by yostashiro — CHANGES_REQUESTED** [review]
Please add comments to briefly explain the background/reason of the adjusted logic.

**Comment by yostashiro** [comment]
@TomohiroAbe-Quartile Just split some chunks into methods as the main method was getting bloated.

**Review by yostashiro — APPROVED** [review]

**Comment by TomohiroAbe-Quartile** [comment]
@yostashiro 
Functional: Worked as expected.
Code: LGTM.

I'll check readme and xml comment-out, and merge this PR.
